moon/controller.py

```diff
@@ -447,25 +522,58 @@ def get_angle_diff(self, destination, direction=1):
 
         return normalizeAnglePIPI(angle_diff)
 
-    def go_to_location(self, pos, speed, offset=0.0, full_turn=False, timeout=None):
+    def go_to_location(self, pos, desired_speed, offset=0.0, full_turn=False, timeout=None, with_stop=True, tolerance=1.0):
         # speed: + forward, - backward
         # offset: stop before (-) or past (+) the actual destination (e.g., to keep the destination in front of the robot)
-        print(self.sim_time, "go_to_location [%.1f,%.1f] (speed: %.1f)" % (pos[0], pos[1], math.copysign(self.max_speed, speed)))
+        print(self.sim_time, self.robot_name, "go_to_location [%.1f,%.1f] (speed: %.1f)" % (pos[0], pos[1], math.copysign(self.max_speed, desired_speed)))
 
+        dist = distance(pos, self.xyz)+offset
         if timeout is None:
-            dist = distance(pos, self.xyz)
             timeout = timedelta(seconds=(10 if full_turn else 0) + 2*dist / self.max_speed) # 30m should take at most 60 seconds, unless turning heavily, then add the amount of time it takes to turn around
 
         start_time = self.sim_time
 
         # while we are further than 1m but still following the right direction (angle diff < 90deg)
-        angle_diff = self.get_angle_diff(pos,speed)
-        while distance(pos, self.xyz)+offset > 1 and (full_turn or abs(angle_diff) < math.pi/2):
-            angle_diff = self.get_angle_diff(pos,speed)
+        angle_diff = self.get_angle_diff(pos,desired_speed)
+        while distance(pos, self.xyz)+offset > tolerance and (full_turn or abs(angle_diff) < math.pi/2):
+            angle_diff = self.get_angle_diff(pos,desired_speed)
             dist = distance(pos, self.xyz)+offset # do not turn just before arrival
-            if angle_diff > 0.1 and dist > 3:
+
+
+            speed = desired_speed if dist > 4 or not with_stop else 0.6 * desired_speed
+
+            if speed > 0 and dist > 5 and (self.scan_distance_to_obstacle < 4000 or (self.avoidance_start is not None and self.sim_time - self.avoidance_start < timedelta(milliseconds=2000))):
+                # do not avoid obstacles when real close
+                if self.scan_distance_to_obstacle < 4000:
+                    self.avoidance_start = self.sim_time
+                    if self.avoidance_turn is None:
+                        if self.scan_avg_distance_left < 0.8 * self.scan_avg_distance_right:
+                            self.avoidance_turn = -AVOID_RADIUS
+                        elif  self.scan_avg_distance_left * 0.8 > self.scan_avg_distance_right:
+                            self.avoidance_turn = AVOID_RADIUS
+                        else:
+                            self.avoidance_turn = AVOID_RADIUS if getrandbits(1) == 0 else -AVOID_RADIUS
+                    turn = self.avoidance_turn
+                    if (self.debug):
```

Comment:
for this would be more reasonable to use some `logging` (INFO, DEBUG levels), but ... one day ...